Enable polymorphism

nakamorichi opened the pull request:
...by merging members of "allOf" into a single properties object and a single requirements array.

Fixes https://github.com/simonguest/swagger-mongoose/issues/29

Commit:
Enable polymorphism by merging members of "allOf" into a single properties object and a single requirements array.

Commit:
export swaggerJSON

Commit:
Fix export

Commit:
Fix typo

briananderson1222 commented:
@Kitanotori Thanks for the contributions here. Are there any unit tests that you can add to ensure that this functionality doesn't undergo regression in the future?

briananderson1222 commented:
If not, I plan to go back and add some tests for the `additionalProperties` functionality I added and might be able to take a look at adding tests here. I think with those two change it may warrant a new release.

nakamorichi commented:
Unfortunately no time to write tests currently.

celalo commented:
This does not handle `$ref` only members of `allOf`, does it? Example: 
```
"definitions": {
        "dealer_output": {
            "title": "Dealer Output",
            "allOf": [
                {
                    "$ref": "#/definitions/dealer_input"
                },
                {
                    "type": "object",
                    "properties": {
                        "_id": {
                            "type": "string",
                            "format": "uuid"
                        }
                    },
                    "required": [
                        "_id"
                    ]
                }
            ]
        }
.
.
.
.
.
```

celalo committed:
handling $ref only members of polymorphic defintions

celalo commented:
This is what I came up with to solve the problem I mentioned above: https://github.com/celalo/swagger-mongoose/commit/8c4370ad604a0745dcc10d318556f07601ec65f9

Commit:
mongodb and mongoose should be peer dependencies

Commit:
mongodb and mongoose should be peer dependencies

Commit:
Remove mongodb dependency as unnecessary (mongoose already depends on it), add mistakenly removed mongoose dependency back to devDependencies

Commit:
Remove mongodb dependency as unnecessary (mongoose already depends on it), add mistakenly removed mongoose dependency back to devDependencies

nakamorichi committed:
async => ^2.0.0, mockgoose => ^7.0.0, remove mongoose dev dep as unnecessary and add necessary peer dep

nakamorichi committed:
Update deps

nakamorichi committed:
Correct some deprecated syntax, set mongoose to use ES6 Promise in tests

nakamorichi committed:
Correct some deprecated syntax, set mongoose to use ES6 Promise in tests

nakamorichi committed:
Add definitions containing "allOf' syntax, and a basic test case

nakamorichi committed:
Merge branch 'complex-defs' into process-allof

# Conflicts:
#	package.json

briananderson1222 commented:
I just saw this! Thanks for pushing forward on these updates! From what I can tell the Travis job is failing due to some nested dependencies that no longer support node 0.11. Feel free to update to 4+ in the travis.yml and see if we can get the tests to pass here.

nakamorichi committed:
Upgrade Node version in Travis to the most recent LTS (i.e. 6)

nakamorichi committed:
Upgrade Node version in Travis to the most recent LTS (i.e. 6)

nakamorichi committed:
Merge branch 'complex-defs' of github.com:Kitanotori/swagger-mongoose into process-allof

nakamorichi committed:
Increase mocha's timeout limit to 10s

nakamorichi committed:
Merge branch 'complex-defs' of github.com:Kitanotori/swagger-mongoose into process-allof

nakamorichi commented:
@briananderson1222 I fixed Travis's and Mocha's settings (timeout limit was too low). Now only the test I added, containing example of allOf definition, fails.

Commit:
‘date’ and ‘double’ are not valid types

Commit:
Fix type handling:

- date and date-time were not processed correctly when their type was correctly specified as ‘string’
- date, dateTime, password, double, float, long are not valid types

Commit:
Increase Mocha’s timeout limit because of Mocha’s slow initialization

Commit:
Merge branch 'update-deps-and-travis' into process-allof

# Conflicts:
#	package.json
#	test/index.js

Commit:
Fix type handling: all ‘integer’s and ‘number’s as Number, all ‘string’s except ‘date’ and ‘date-time’ as String

LGLC commented on lib/index.js:
```diff
@@ -208,11 +203,53 @@ var processRef = function (property, objectName, props, key, required) {
+
 var getSchema = function (objectName, fullObject) {
   var props = {};
   var required = fullObject.required || [];
```
Line 247 (`var required = fullObject.required || [];`) should be moved to line 251 (i.e. after processPolymorphic runs if the definition is polymorphic) otherwise we're not adding in the merged required fields.
